### PR TITLE
Download and Kinetics 400/600/700 Datasets

### PR DESCRIPTION
**Pull request by bjuncek**
Addresses #3452.

### CODE REVIEW

**Comment by bjuncek** [comment]
@pmeier is there a nice/neat way to read in CSV annotation file and have it be queryable  _without_ using pandas?

File: torchvision/datasets/kinetics.py [hunk 1]
```diff
@@ -35,29 +48,92 @@ class Kinetics400(VisionDataset):
...
+    _ANNOTATION = {
+        400: "https://s3.amazonaws.com/kinetics/400/annotations/{split}.csv",
+        600: "https://s3.amazonaws.com/kinetics/600/annotations/{split}.txt",
+        700: "https://s3.amazonaws.com/kinetics/700_2020/annotations/{split}.csv",
+    }
+
+    def __init__(
+        self,
+        root: str,
+        n_classes: int = 400,
```
Comment (pmeier) [inline torchvision/datasets/kinetics.py:90]: Although naturally an `int` would be correct, this also implies that I could set an arbitrary number. Thus I would go with `"400"`, `"600"`, and `"700"`.

```suggestion
        num_classes: str = "400",
```

File: torchvision/datasets/kinetics.py [hunk 2]
```diff
@@ -35,29 +48,92 @@ class Kinetics400(VisionDataset):
...
+    ) -> None:
+
+        # TODO: support test
+        assert split in ["train", "val"]
+        assert n_classes in [400, 700]
+        self.n_classes = n_classes
+        self.extensions = extensions
+        self._num_download_workers = _num_download_workers
+
+        # set up self.root and self.split
```
Comment (pmeier) [inline torchvision/datasets/kinetics.py:116]: This comment is redundant if the method invocation is expressive which it is.

File: torchvision/datasets/kinetics.py [hunk 3]
```diff
@@ -35,29 +48,92 @@ class Kinetics400(VisionDataset):
...
+        self._set_up_paths(root, split)
 
+        # load annotation files
```
Comment (pmeier) [inline torchvision/datasets/kinetics.py:119]: Same as above.

File: torchvision/datasets/kinetics.py [hunk 4]
```diff
@@ -74,6 +150,106 @@ def __init__(self, root, frames_per_clip, step_between_clips=1, frame_rate=None,
...
+                    )
+                    os.makedirs(os.path.join(self.root, label), exist_ok=True)
+                    os.replace(
+                        os.path.join(self.root, file),
+                        os.path.join(self.root, label, file),
+                    )
+                except:
+                    warnings.warn(
+                        f"Unexpected error while processing {ytid}:", sys.exc_info()[0]
+                    )
```
Comment (pmeier) [inline torchvision/datasets/kinetics.py:251]:
```suggestion
                except Exception as error:
                    warnings.warn(
                        f"Unexpected error while processing {ytid}: {error}"
                    )
```

File: torchvision/datasets/kinetics.py [hunk 5]
```diff
@@ -74,6 +150,106 @@ def __init__(self, root, frames_per_clip, step_between_clips=1, frame_rate=None,
...
+        )
+        kinetics_dir, _ = path.split(self.root)
+        tar_path = path.join(kinetics_dir, "tars")
+
+        if self._num_download_workers < 2:
+            for line in file_url:
+                line = str(line.decode("utf-8")).replace("\n", "")
+                dl_wrap(tar_path, self.root, line)
+        else:
+            part = partial(_dl_wrap, tar_path, self.root)
```
Comment (pmeier) [inline torchvision/datasets/kinetics.py:211]: Whats wrong with doing this directly with `download_and_extract_archive`?

File: torchvision/datasets/kinetics.py [hunk 6]
```diff
@@ -35,29 +48,92 @@ class Kinetics400(VisionDataset):
...
+        # set up self.root and self.split
+        self._set_up_paths(root, split)
 
+        # load annotation files
+        if annotation_path is not None:
+            self.annotations = pd.read_csv(annotation_path)
+        else:
+            self.annotations = pd.read_csv(
+                self._ANNOTATION[self.n_classes].format(split=self.split)
+            )
```
Comment (pmeier) [inline torchvision/datasets/kinetics.py:125]: Re https://github.com/pytorch/vision/pull/3680#issuecomment-821678875: The annotation files are structured like this:

```
label,youtube_id,time_start,time_end,split
clay pottery making,---0dWlqevI,19,29,train
...
```

It seems you are only using the fields `label` and `youtube_id`.

1. Is that correct? Shouldn't we respect `time_start` and `time_end`?
2. If we really only need two fields per row, we probably go the easy way:
    
    ```python
    import csv
    
    with open("annotations.csv") as fh:
        label, youtube_id = zip(*[row[:2] for row in csv.reader(fh, delimiter=",")])
    ```

    In case we also need access to the other fields, a [`DictReader`]() would probably a better choice. 

    In any case this will most likely be slower than the `pandas` implementation. We just recently removed the last `pandas` dependency in #3656. Before we reintroduce it here, we should profile this and see if it is a bottleneck.

File: torchvision/datasets/kinetics.py [hunk 7]
```diff
@@ -74,6 +150,106 @@ def __init__(self, root, frames_per_clip, step_between_clips=1, frame_rate=None,
...
+                f"delete the directory."
+            )
+
+        file_url = urllib.request.urlopen(
+            self._FILES[self.n_classes].format(split=self.split)
+        )
+        kinetics_dir, _ = path.split(self.root)
+        tar_path = path.join(kinetics_dir, "tars")
+
+        if self._num_download_workers < 2:
```
Comment (pmeier) [inline torchvision/datasets/kinetics.py:206]: Maybe turn this block around to make the intent more clear:

```suggestion
        if self._num_download_workers > 1:
```

File: torchvision/datasets/kinetics.py [hunk 8]
```diff
@@ -74,6 +150,106 @@ def __init__(self, root, frames_per_clip, step_between_clips=1, frame_rate=None,
...
+            for line in file_url:
+                line = str(line.decode("utf-8")).replace("\n", "")
+                dl_wrap(tar_path, self.root, line)
+        else:
+            part = partial(_dl_wrap, tar_path, self.root)
+            lines = [str(line.decode("utf-8")).replace("\n", "") for line in file_url]
+            poolproc = Pool(self._num_download_workers)
+            poolproc.map(part, lines)
+        return 0
```
Comment (pmeier) [inline torchvision/datasets/kinetics.py:215]: ?

**Review by pmeier — CHANGES_REQUESTED** [review]
Hey Bruno, thanks for the PR! I did an initial pass. Comments below.

File: torchvision/datasets/kinetics.py (hunk 2 above)
Comment (bjuncek) [reply torchvision/datasets/kinetics.py:116]: addressed

File: torchvision/datasets/kinetics.py (hunk 6 above)
Comment (bjuncek) [reply torchvision/datasets/kinetics.py:125]: In theory yes, but the videos provided are already cut to the respective lens so we don't need them.

I'll rewrite this using `csv` then to avoid adding the additional dependencies.

File: torchvision/datasets/kinetics.py (hunk 5 above)
Comment (bjuncek) [reply torchvision/datasets/kinetics.py:211]: I was having some isseus with `partial` and that; maybe we can discuss offline.

File: torchvision/datasets/kinetics.py (hunk 6 above)
Comment (pmeier) [reply torchvision/datasets/kinetics.py:125]: > In theory yes, but the videos provided are already cut to the respective lens so we don't need them.

Ah ok, so this is just the information how the file was obtained not instructions how to obtain it?

File: torchvision/datasets/kinetics.py (hunk 5 above)
Comment (bjuncek) [reply torchvision/datasets/kinetics.py:211]: partial can only be applied to the leftmost argument, so not really

File: torchvision/datasets/kinetics.py (hunk 6 above)
Comment (bjuncek) [reply torchvision/datasets/kinetics.py:125]: I'll rewrite this with CSV then;
we don't need `time_start` and `time_end` as the official videos are already pre-trimmed.

**Comment by bjuncek** [comment]
@pmeier I've fixed all the issues you mentioned, plus removed pandas dependency. Can you please check if all looks well. 
Also, additionally, should we add a download test?

**Review by pmeier — CHANGES_REQUESTED** [review]
Mostly style related comments left. 

> Also, additionally, should we add a download test?

Yes, please. If you need help to set them up, please let me know. Happy to help. 

Plus, since this adds new parameters (`n_classes` and `split`), we should also include them in our dataset test

https://github.com/pytorch/vision/blob/d7ad62ff83bb329b7e5f2725bfa84e1266febc55/test/test_datasets.py#L1067

You can do so by adding a `ADDITIONAL_CONFIGS` class variable like

```python
ADDITIONAL_CONFIGS = datasets_utils.combinations_grid(
        n_classes=("400", "600", "700"), split=("train", "val")
    )
```

File: torchvision/datasets/kinetics.py [hunk 9]
```diff
@@ -35,29 +48,85 @@ class Kinetics400(VisionDataset):
                 └── class2
                     ├── clipx.avi
                     └── ...
-
+            If the split is not defined, it is appended using the split argument.
+        n_classes (int): select between Kinetics-400, Kinetics-600, and Kinetics-700
```
[the remark this answers is not in the document]
Comment (bjuncek) [reply torchvision/datasets/kinetics.py:52]: makes sense, done

File: torchvision/datasets/kinetics.py [hunk 10]
```diff
@@ -20,11 +35,9 @@ class Kinetics400(VisionDataset):
     Note that we drop clips which do not have exactly ``frames_per_clip`` elements, so not all
     frames in a video might be present.
 
-    Internally, it uses a VideoClips object to handle clip creation.
-
     Args:
-        root (string): Root directory of the Kinetics-400 Dataset. Should be structured as follows:
-
+        root (string): Root directory of the (split of the) Kinetics Dataset.
+            Directory should be structured as follows:
```
[the remark this answers is not in the document]
Comment (bjuncek) [reply torchvision/datasets/kinetics.py:40]: done

File: torchvision/datasets/kinetics.py [hunk 11]
```diff
@@ -89,3 +245,90 @@ def __getitem__(self, idx):
...
+    Returns:
+        tuple: A 3-tuple with the following entries:
+
+            - video (Tensor[T, H, W, C]): the `T` video frames
+            - audio(Tensor[K, L]): the audio frames, where `K` is the number of channels
+              and `L` is the number of points
+            - label (int): class of the video clip
+    """
+
+    def __init__(
```
[the remark this answers is not in the document]
Comment (bjuncek) [reply torchvision/datasets/kinetics.py:293]: I don't think we should do it as people might be relying on this dataset as is (without additional kwargs).

File: torchvision/datasets/kinetics.py [hunk 12]
```diff
@@ -89,3 +245,90 @@ def __getitem__(self, idx):
...
+        self,
+        root,
+        frames_per_clip,
+        step_between_clips=1,
+        frame_rate=None,
+        extensions=("avi",),
```
[the remark this answers is not in the document]
Comment (bjuncek) [reply torchvision/datasets/kinetics.py:299]: I think we should add this regardless of the back port, lest someone downloads the data manually and runs into an error. ".avi" was done I think purely because that's how the most popular download script encoded the data.

File: torchvision/datasets/kinetics.py [hunk 13]
```diff
@@ -74,6 +143,93 @@ def __init__(self, root, frames_per_clip, step_between_clips=1, frame_rate=None,
...
+
+    def _download_videos(self) -> None:
+        """download tarballs containing the video to
+        "tars" folder and extract them into the _split_ folder
+        where split is one of the official dataset splits.
+
+        Raises:
+            RuntimeError: if download folder exists, break to prevent
+              downloading entire dataset again.
+        """
```
[the remark this answers is not in the document]
Comment (bjuncek) [reply torchvision/datasets/kinetics.py:165]: done

File: torchvision/datasets/kinetics.py [hunk 14]
```diff
@@ -74,6 +143,93 @@ def __init__(self, root, frames_per_clip, step_between_clips=1, frame_rate=None,
...
+                    .replace("'", "")
+                    .replace("(", "")
+                    .replace(")", "")
+                )
+                os.makedirs(os.path.join(self.root, label), exist_ok=True)
+                existing_file = os.path.join(self.root, f)
+                if os.path.isfile(existing_file):
+                    os.replace(
+                        existing_file, os.path.join(self.root, label, f),
+                    )
```
[the remark this answers is not in the document]
Comment (bjuncek) [reply torchvision/datasets/kinetics.py:227]: yes; however if someone uses the outdated annotation files (not sure why they'd want to, but they can) they'd run into this.

I can remove it and have a try/catch with raising a warning as a better way of handling this issue?

File: torchvision/datasets/kinetics.py [hunk 15]
```diff
@@ -74,6 +143,93 @@ def __init__(self, root, frames_per_clip, step_between_clips=1, frame_rate=None,
...
+    def _make_ds_structure(self):
+        """move videos from
+        root/
+            ├── clip1.avi
+            ├── clip2.avi
+
+        to the correct format as described below:
+        root/
+            ├── class1
+            │   ├── clip1.avi
```
[the remark this answers is not in the document]
Comment (bjuncek) [reply torchvision/datasets/kinetics.py:203]: So, we've used `.avi` as the most prominent Kinetics download code encoded it as `.avi`. The current files are all `.mp4` so I propose to use all .mp4 in the new doctoring, but keep a mix in the old one as a legacy support.

File: torchvision/datasets/kinetics.py (hunk 14 above)
[the remark this answers is not in the document]
Comment (bjuncek) [reply torchvision/datasets/kinetics.py:225]: actually, even with their annotation files, some videos do not exist so this is very much needed

File: test/test_datasets_download.py [hunk 16]
```diff
@@ -391,6 +391,25 @@ def widerface():
     )
 
 
+def kinetics():
+    return itertools.chain(
+        *[
+            collect_download_configs(
+                lambda: datasets.Kinetics(
+                    path.join(ROOT, "Kinetics", split),
```
Comment (pmeier) [inline test/test_datasets_download.py:399]: @bjuncek I don't see any reason why the user needs to manually set the last part of the path to the correct split. This should be handled internally.

```suggestion
                    ROOT,
```

Comment (bjuncek) [reply test/test_datasets_download.py]: makes sense; it was done to avoid breaking BC with the API, but we can do this more elegantly.
Please check my subsequent commit for that.

File: test/test_datasets_download.py [hunk 17]
```diff
@@ -392,6 +392,22 @@ def widerface():
...
+                    frames_per_clip=1,
+                    num_classes=num_classes,
+                    split=split,
+                    download=True,
+                ),
+                name=f"Kinetics, {num_classes}, {split}",
+                file="kinetics",
+            )
+            for num_classes, split in itertools.product(("400", "600", "700"), ("train", "val"))
+
```
Comment (pmeier) [inline test/test_datasets_download.py:410]:
```suggestion
            for num_classes, split in itertools.product(("400", "600", "700"), ("train", "val"))
        ]
    )
```

File: test/test_datasets.py [hunk 18]
```diff
@@ -944,14 +943,44 @@ def test_not_found_or_corrupted(self):
...
+        return root, 1
+
     def inject_fake_data(self, tmpdir, config):
         classes = ("Abseiling", "Zumba")
         num_videos_per_class = 2
 
         digits = string.ascii_letters + string.digits + "-_"
+        # note: train is here hardcoded by default bc we expect the user to supply it,
+        # but that requirement have changed in subsequent version of the dataset
+        tmpdir = pathlib.Path(tmpdir) / "train"
```
Comment (pmeier) [inline test/test_datasets.py:983]: We should aim for BC. That means we shouldn't need to change anything here and the tests should pass.

File: torchvision/datasets/kinetics.py [hunk 19]
```diff
@@ -20,44 +34,103 @@ class Kinetics400(VisionDataset):
...
+        split (str): split of the dataset to consider; currently supports ["train", "val"]
+        frame_rate (float): If not None, interpolate different frame rate for each clip.
         frames_per_clip (int): number of frames in a clip
         step_between_clips (int): number of frames between each clip
+        annotation_path (str): path to official Kinetics annotation file.
         transform (callable, optional): A function/transform that  takes in a TxHxWxC video
             and returns a transformed version.
+        download (bool): Download the official version of the dataset to root folder.
+        num_workers (int): Use multiple workers for VideoClips creation
+        _num_download_workers (int): Use multiprocessing in order to speed up download.
```
Comment (pmeier) [inline torchvision/datasets/kinetics.py:63]: Either make this public (remove the underscore) or remove it from the docstring. No hard opinion here, but since the download is quite lengthy I would make it public. Otherwise no one will know about this unless the dive into the source.

File: torchvision/datasets/kinetics.py [hunk 20]
```diff
@@ -20,44 +34,103 @@ class Kinetics400(VisionDataset):
...
 
-            - video (Tensor[T, H, W, C]): the `T` video frames
+            - video (Tensor[T, H, W, C]): the `T` video frames in torch.uint8 tensor
             - audio(Tensor[K, L]): the audio frames, where `K` is the number of channels
-              and `L` is the number of points
+              and `L` is the number of points in torch.float tensor
             - label (int): class of the video clip
+
+    Raises:
+        RuntimeError: If ``download is True`` and the image archive is already extracted.
```
Comment (pmeier) [inline torchvision/datasets/kinetics.py:74]:
```suggestion
        RuntimeError: If ``download is True`` and the video archives are already extracted.
```

File: torchvision/datasets/kinetics.py [hunk 21]
```diff
@@ -20,44 +34,103 @@ class Kinetics400(VisionDataset):
...
+        _precomputed_metadata=None,
+        _num_download_workers=1,
+        _video_width=0,
+        _video_height=0,
+        _video_min_dimension=0,
+        _audio_samples=0,
+        _audio_channels=0,
+    ) -> None:
+
+        # TODO: support test
```
Comment (pmeier) [inline torchvision/datasets/kinetics.py:110]: Go for it in a follow-up PR if you like!

File: torchvision/datasets/kinetics.py [hunk 22]
```diff
@@ -20,44 +34,103 @@ class Kinetics400(VisionDataset):
...
+            self.annotations = annotation_path
+
+        if download:
+            self.download_and_process_videos()
+
+        super().__init__(self.root)
 
         self.classes, class_to_idx = find_classes(self.root)
         self.samples = make_dataset(self.root, class_to_idx, extensions, is_valid_file=None)
+
```
Comment (pmeier) [inline torchvision/datasets/kinetics.py:133]:
```suggestion
```

File: torchvision/datasets/kinetics.py [hunk 23]
```diff
@@ -89,3 +247,92 @@ def __getitem__(self, idx):
...
+        frame_rate=None,
+        extensions=("avi", "mp4"),
+        transform=None,
+        _precomputed_metadata=None,
+        num_workers=1,
+        _video_width=0,
+        _video_height=0,
+        _video_min_dimension=0,
+        _audio_samples=0,
+        _audio_channels=0,
```
Comment (pmeier) [inline torchvision/datasets/kinetics.py:313]: The purpose of `**kwargs` is to not repeat all keyword arguments that should be the same. 

```suggestion
```

File: torchvision/datasets/kinetics.py [hunk 24]
```diff
@@ -89,3 +247,92 @@ def __getitem__(self, idx):
...
+            extensions=extensions,
+            transform=transform,
+            _precomputed_metadata=_precomputed_metadata,
+            num_workers=num_workers,
+            _video_width=_video_width,
+            _video_height=_video_height,
+            _video_min_dimension=_video_min_dimension,
+            _audio_channels=_audio_channels,
+            _audio_samples=_audio_samples,
+            download=False,
```
Comment (pmeier) [inline torchvision/datasets/kinetics.py:337]: See above

```suggestion
            download=False,
            **kwargs,
```

File: torchvision/datasets/kinetics.py [hunk 25]
```diff
@@ -74,6 +147,91 @@ def __init__(self, root, frames_per_clip, step_between_clips=1, frame_rate=None,
...
+        # check that the assignment was made properly
+        kinetics_dir, split = path.split(self.root)
+        assert split == self.split, 'File folder assignment not done properly'
+        tar_path = path.join(kinetics_dir, "tars")
+        annotation_path = path.join(kinetics_dir, "annotations")
+        file_list_path = path.join(kinetics_dir, "files")
+
+        split_url = self._TAR_URLS[self.num_classes].format(split=self.split)
+        split_url_filepath = path.join(file_list_path, path.basename(split_url))
+        if not path.isfile(split_url_filepath):
```
Comment (pmeier) [inline torchvision/datasets/kinetics.py:182]: Use `utils.check_integrity` here. Makes it easier to add md5sum checking later.

File: torchvision/datasets/kinetics.py [hunk 26]
```diff
@@ -74,6 +147,91 @@ def __init__(self, root, frames_per_clip, step_between_clips=1, frame_rate=None,
...
+            download_url(split_url, file_list_path)
+        list_video_urls = open(split_url_filepath, "r")
+
+        if not path.isfile(path.join(annotation_path, f"{self.split}.csv")):
```
Comment (pmeier) [inline torchvision/datasets/kinetics.py:186]: Same as above.

File: torchvision/datasets/kinetics.py [hunk 27]
```diff
@@ -20,44 +34,103 @@ class Kinetics400(VisionDataset):
...
+        self._num_download_workers = _num_download_workers
+
+        if path.basename(root) != split:
+            self.root = path.join(root, split)
+        else:
+            self.root = root
+        self.split = split
+
+        if annotation_path is not None:
+            self.annotations = annotation_path
```
Comment (pmeier) [inline torchvision/datasets/kinetics.py:124]: It is not a good idea to have an attribute that only exist for a specific parameter combination. Given that we only use the annotations in the download / restructuring we should be able to remove it here and only use it locally.

File: torchvision/datasets/kinetics.py [hunk 28]
```diff
@@ -20,44 +34,103 @@ class Kinetics400(VisionDataset):
...
+                │   ├──  class2
+                │   │   ├──   clipx.mp4
+                │   │    └── ...
+            Split is appended using the split argument.
+        num_classes (int): select between Kinetics-400, Kinetics-600, and Kinetics-700
+        split (str): split of the dataset to consider; currently supports ["train", "val"]
+        frame_rate (float): If not None, interpolate different frame rate for each clip.
         frames_per_clip (int): number of frames in a clip
         step_between_clips (int): number of frames between each clip
+        annotation_path (str): path to official Kinetics annotation file.
```
Comment (pmeier) [inline torchvision/datasets/kinetics.py:58]: I don't think we need this as parameter. What would be the use case?

File: torchvision/datasets/kinetics.py [hunk 29]
```diff
@@ -89,3 +247,92 @@ def __getitem__(self, idx):
...
+        _video_min_dimension=0,
+        _audio_samples=0,
+        _audio_channels=0,
+        **kwargs
+    ):
+        warnings.warn(
+            "Kinetics400 is deprecated and will be removed in a future release."
+            "It was replaced by Kinetics(..., num_classes=\"400\").")
+
+        kinetics_dir, split = path.split(root)
```
Comment (pmeier) [inline torchvision/datasets/kinetics.py:320]: I'm quite unhappy about this because it might break BC. This will only work if a user uses the `root/split` way to store the data. For example:

Instead of changing `root` in the new `Kinetics` class we keep it the same way it was passed to us and create a new attribute `split_folder`. Additionally we add private attribute `_legacy_structure=False`. It will be set to `True` by the old `Kinectics400` and in this case our `split_folder` simply points to `root`. We could allow an undocumented `split="unknown"` in case a user looks at the `split` attribute.

File: test/test_datasets.py [hunk 30]
```diff
@@ -841,7 +841,6 @@ def _create_annotation_files(self, root, video_files, fold, train):
         other_annotations.remove(current_annotation)
         for name in other_annotations:
             self._create_annotation_file(root, name, other_videos)
-
```
Comment (pmeier) [inline test/test_datasets.py:844]: Please revert as this is unrelated.

**Review by pmeier — CHANGES_REQUESTED** [review]
Looking good Bruno! One major and a few minor things left.

File: torchvision/datasets/kinetics.py (hunk 28 above)
Comment (bjuncek) [reply torchvision/datasets/kinetics.py:58]: if someone decided to use their own additional annotations (e.g. for action detection) or different version of the annotations (legacy ones with old videos)

File: torchvision/datasets/kinetics.py (hunk 27 above)
Comment (bjuncek) [reply torchvision/datasets/kinetics.py:124]: actually, yeah, makes sense to remove this. Will do

File: torchvision/datasets/kinetics.py (hunk 28 above)
Comment (bjuncek) [reply torchvision/datasets/kinetics.py:58]: actually, yeah, makes sense to remove this. Will do

File: torchvision/datasets/kinetics.py [hunk 31]
```diff
@@ -20,44 +34,104 @@ class Kinetics400(VisionDataset):
...
+        _video_min_dimension=0,
+        _audio_samples=0,
+        _audio_channels=0,
+        **kwargs
+    ) -> None:
+
+        # TODO: support test
+        verify_str_arg(split, arg="split", valid_values=['train', 'val', 'unknown'])
+        verify_str_arg(num_classes, arg="num_classes", valid_values=["400", "600", "700"])
+        self.num_classes = num_classes
```
Comment (pmeier) [inline torchvision/datasets/kinetics.py:112]: `verify_str_arg` returns the checked value for convenience

```suggestion
        self.num_classes = verify_str_arg(num_classes, arg="num_classes", valid_values=["400", "600", "700"])
```

File: torchvision/datasets/kinetics.py [hunk 32]
```diff
@@ -20,44 +34,104 @@ class Kinetics400(VisionDataset):
...
+        if _use_legacy_structure:
+            print("Using legacy structure")
+            self.root = root
+            self.split_folder = root
+            self.split = "unknown"
+            assert download == False, "Cannot download the videos using legacy_structure."
+        else:
+            self.root = root
+            self.split_folder = path.join(root, split)
+            self.split = split
```
Comment (pmeier) [inline torchvision/datasets/kinetics.py:126]:
```suggestion
            self.split = verify_str_arg(split, arg="split", valid_values=["train", "val"])
```

File: torchvision/datasets/kinetics.py [hunk 33]
```diff
@@ -20,44 +34,104 @@ class Kinetics400(VisionDataset):
...
+        self.extensions = extensions
+        self.num_download_workers = num_download_workers
+
+        _use_legacy_structure = kwargs.get('_use_legacy_structure', False)
+        if _use_legacy_structure:
+            print("Using legacy structure")
+            self.root = root
+            self.split_folder = root
+            self.split = "unknown"
+            assert download == False, "Cannot download the videos using legacy_structure."
```
Comment (pmeier) [inline torchvision/datasets/kinetics.py:122]:
```suggestion
            assert not download, "Cannot download the videos using legacy_structure."
```

File: torchvision/datasets/kinetics.py [hunk 34]
```diff
@@ -89,3 +246,68 @@ def __getitem__(self, idx):
...
+
+    def __init__(
+        self,
+        root,
+        frames_per_clip,
+        **kwargs
+    ):
+        warnings.warn(
+            "Kinetics400 is deprecated and will be removed in a future release."
+            "It was replaced by Kinetics(..., num_classes=\"400\").")
```
Comment (pmeier) [inline torchvision/datasets/kinetics.py:306]: Instead of allowing the new parameters (`split`, `num_classes`, `download`, `num_download_workers`) here and only fail when we reach the super class, we could do something like this:

```suggestion
        num_classes = None,
        split = None,
        download = None,
        num_download_workers = None,
        **kwargs
    ):
        warnings.warn(
            "Kinetics400 is deprecated and will be removed in a future release."
            "It was replaced by Kinetics(..., num_classes=\"400\").")
        if any(value is not None for value in (num_classes, split, download, num_download_workers)):
            raise RuntimeError(
                "Usage of 'num_classes', 'split', 'download', or 'num_download_workers' is not supported in Kinetics400. "
                "Please use Kinetics instead."
            )
```

File: torchvision/datasets/kinetics.py [hunk 35]
```diff
@@ -20,44 +34,104 @@ class Kinetics400(VisionDataset):
...
+            assert download == False, "Cannot download the videos using legacy_structure."
+        else:
+            self.root = root
+            self.split_folder = path.join(root, split)
+            self.split = split
+
+        if download:
+            self.download_and_process_videos()
+
+        super().__init__(self.split_folder)
```
Comment (pmeier) [inline torchvision/datasets/kinetics.py:131]: We should pass `root` here so `root == Kinectics(root, ...).root`. Internally we can then use `split_folder` for everything.

File: torchvision/datasets/kinetics.py [hunk 36]
```diff
@@ -74,6 +148,89 @@ def __init__(self, root, frames_per_clip, step_between_clips=1, frame_rate=None,
...
+        split_folder/
+            ├── class1
+            │   ├── clip1.avi
+
+        """
+        file_tmp = "{ytid}_{start:06}_{end:06}.mp4"
+        with open(self.annotations) as csvfile:
+            reader = csv.DictReader(csvfile)
+            for row in reader:
+                f = file_tmp.format(
```
Comment (pmeier) [inline torchvision/datasets/kinetics.py:215]: Nit: tmp is not expressive.

```suggestion
        file_fmtstr = "{ytid}_{start:06}_{end:06}.mp4"
        with open(self.annotations) as csvfile:
            reader = csv.DictReader(csvfile)
            for row in reader:
                f = file_fmtstr.format(
```

File: torchvision/datasets/kinetics.py [hunk 37]
```diff
@@ -74,6 +148,89 @@ def __init__(self, root, frames_per_clip, step_between_clips=1, frame_rate=None,
...
+                    .replace("'", "")
+                    .replace("(", "")
+                    .replace(")", "")
+                )
+                os.makedirs(path.join(self.split_folder, label), exist_ok=True)
+                existing_file = path.join(self.split_folder, f)
+                if path.isfile(existing_file):
+                    os.replace(
+                        existing_file, path.join(self.split_folder, label, f),
+                    )
```
Comment (pmeier) [inline torchvision/datasets/kinetics.py:232]: I remember asking myself this before, but I can't find the comment if I added one: how can a file be pre-existing here? Given that we always start from a clean folder, the only possibility is a duplicate entry in the annotations file. If this happens, an error is warranted, no?

File: torchvision/datasets/kinetics.py [hunk 38]
```diff
@@ -74,6 +148,89 @@ def __init__(self, root, frames_per_clip, step_between_clips=1, frame_rate=None,
...
+
+        split_url = self._TAR_URLS[self.num_classes].format(split=self.split)
+        split_url_filepath = path.join(file_list_path, path.basename(split_url))
+        if not check_integrity(split_url_filepath):
+            download_url(split_url, file_list_path)
+        list_video_urls = open(split_url_filepath, "r")
+
+        if not check_integrity(path.join(annotation_path, f"{self.split}.csv")):
+            download_url(self._ANNOTATION_URLS[self.num_classes].format(split=self.split), annotation_path)
+        self.annotations = path.join(annotation_path, f"{self.split}.csv")
```
Comment (pmeier) [inline torchvision/datasets/kinetics.py:187]: Instead of setting this to `self` here, you could return it and pass it on to `_make_ds_structure`. Otherwise we still have the same problem as before: the attribute `annotations` only exists when `download=True` was used.

**Review by pmeier — APPROVED** [review]
Hey Bruno, thanks for the awesome work! I have only minor comments left. Afterwards this is good to merge :tada:

File: torchvision/datasets/kinetics.py (hunk 37 above)
Comment (bjuncek) [reply torchvision/datasets/kinetics.py:232]: Actually, what happens is a missing entry; there are several videos missing that still exist in official annotations for their "backwards compatibility". I can raise an error, but we have to have this otherwise it breaks.

**Review by pmeier — APPROVED** [review]
Looks awesome, thanks Bruno!

File: torchvision/datasets/kinetics.py [hunk 39]
```diff
@@ -89,3 +241,77 @@ def __getitem__(self, idx):
...
+
+        frames_per_clip (int): number of frames in a clip
+        step_between_clips (int): number of frames between each clip
+        transform (callable, optional): A function/transform that  takes in a TxHxWxC video
+            and returns a transformed version.
+
+    Returns:
+        tuple: A 3-tuple with the following entries:
+
+            - video (Tensor[T, H, W, C]): the `T` video frames
```
Comment (fmassa) [inline torchvision/datasets/kinetics.py:287]: We should probably not be returning videos as `T, H, W, C` anymore, but instead `T, C, H, W`. I regret having returned a non-standard format for videos in the past.

If we are deprecating the old Kinetics400 dataset, we should take this information into account and do it right this time.

This will also make it easier to use the rest of the video transforms and the new video reader logic

**Review by fmassa — COMMENTED** [review]
This looks good to me, thanks Bruno!

I only have one comment regarding the order of the dimensions for a frame. If we are to be deprecating Kinetics400, I think we might as well change the format to perform a `.transpose()` if we use the legacy format

File: torchvision/datasets/kinetics.py (hunk 39 above)
Comment (pmeier) [reply torchvision/datasets/kinetics.py:287]: Good catch, I wasn't aware of that. @bjuncek we could rename `_use_legacy_structure` to `_legacy` and group the behavior under that.

**Review by fmassa — APPROVED** [review]
Thanks!
